testing: TestRandomSyntaxGeneration panic: timeout

https://teamcity.cockroachdb.com/viewLog.html?buildId=739001&tab=buildResultsDiv&buildTypeId=Cockroach_Nightlies_NightlySuite

```
panic: timeout: "ALTER USER PLACEHOLDER WITH PASSWORD INDEXES". currently executing: map[CREATE DATABASE IF NOT EXISTS ident:1 GRANT ALL ON TABLE ident , COALESCE TO ALTER:1 GRANT ALL ON IFERROR TO ident:1 REVOKE ALL ON SESSION FROM DECIMAL , ROW , SESSION , ident , SUBSTRING:1 ALTER USER PLACEHOLDER WITH PASSWORD INDEXES:1 GRANT ALL ON TABLE TIMESTAMP , BLOB . ident . * , ident TO ident:1 REVOKE ALL ON DATABASE ident , SMALLINT , ident , ident , BOOLEAN FROM TEMPORARY:1 GRANT GRANT , SELECT ON OVERLAY TO CURRENT , REAL:1 REVOKE ALL ON DATABASE COALESCE FROM JSON:1 DROP USER PLACEHOLDER , PLACEHOLDER , PLACEHOLDER:1 GRANT ALL ON WORK TO ident:1 REVOKE CREATE , SELECT ON PASSWORD . * FROM SERIAL8:1 GRANT GRANT ON LIST TO DOUBLE:1 GRANT GRANT , SELECT ON TABLE BY TO ident , ident:1 ALTER USER IF EXISTS ILIKE WITH PASSWORD PLACEHOLDER:1 REVOKE ALL ON START FROM TREAT , COALESCE:1 GRANT ALL ON DATABASE ident , PRECISION , ident TO ident:1 DROP USER IF EXISTS NULLS:1 DROP DATABASE ident:1 ALTER DATABASE ident RENAME TO NUMERIC:1]
```

CC @mjibson

--- a/pkg/internal/client/sender.go
+++ b/pkg/internal/client/sender.go
@@ -151,6 +151,10 @@ type TxnSender interface {
 	// timestamp.
 	CommitTimestamp() hlc.Timestamp
 
+	// CommitTimestampFixed returns true if the commit timestamp has
+	// been fixed to the start timestamp and cannot be pushed forward.
+	CommitTimestampFixed() bool
+
 	// IsSerializablePushAndRefreshNotPossible returns true if the transaction is
 	// serializable, its timestamp has been pushed and there's no chance that
 	// refreshing the read spans will succeed later (thus allowing the transaction
@@ -270,6 +274,11 @@ func (m *MockTransactionalSender) OrigTimestamp() hlc.Timestamp {
 // CommitTimestamp is part of the TxnSender interface.
 func (m *MockTransactionalSender) CommitTimestamp() hlc.Timestamp {
 	return m.txn.OrigTimestamp
+}
+
+// CommitTimestampFixed is part of the TxnSender interface.
+func (m *MockTransactionalSender) CommitTimestampFixed() bool {
+	panic("unimplemented")
 }
 
 // SetFixedTimestamp is part of the TxnSender interface.

--- a/pkg/internal/client/txn.go
+++ b/pkg/internal/client/txn.go
@@ -275,6 +275,12 @@ func (txn *Txn) CommitTimestamp() hlc.Timestamp {
 	return txn.mu.sender.CommitTimestamp()
 }
 
+// CommitTimestampFixed returns true if the commit timestamp has
+// been fixed to the start timestamp and cannot be pushed forward.
+func (txn *Txn) CommitTimestampFixed() bool {
+	return txn.mu.sender.CommitTimestampFixed()
+}
+
 // SetSystemConfigTrigger sets the system db trigger to true on this transaction.
 // This will impact the EndTransactionRequest.
 func (txn *Txn) SetSystemConfigTrigger() error {

--- a/pkg/kv/txn_coord_sender.go
+++ b/pkg/kv/txn_coord_sender.go
@@ -957,6 +957,13 @@ func (tc *TxnCoordSender) CommitTimestamp() hlc.Timestamp {
 	return tc.mu.txn.OrigTimestamp
 }
 
+// CommitTimestampFixed is part of the client.TxnSender interface.
+func (tc *TxnCoordSender) CommitTimestampFixed() bool {
+	tc.mu.Lock()
+	defer tc.mu.Unlock()
+	return tc.mu.txn.OrigTimestampWasObserved
+}
+
 // SetFixedTimestamp is part of the client.TxnSender interface.
 func (tc *TxnCoordSender) SetFixedTimestamp(ctx context.Context, ts hlc.Timestamp) {
 	tc.mu.Lock()

--- a/pkg/sql/backfill.go
+++ b/pkg/sql/backfill.go
@@ -252,11 +252,7 @@ func (sc *SchemaChanger) truncateIndexes(
 				}
 
 				tc := &TableCollection{leaseMgr: sc.leaseMgr}
-				defer func() {
-					if err := tc.releaseTables(ctx, dontBlockForDBCacheUpdate); err != nil {
-						log.Warningf(ctx, "error releasing tables: %s", err)
-					}
-				}()
+				defer tc.releaseTables(ctx)
 				tableDesc, err := sc.getTableVersion(ctx, txn, tc, version)
 				if err != nil {
 					return err
@@ -469,11 +465,7 @@ func (sc *SchemaChanger) distBackfill(
 
 			tc := &TableCollection{leaseMgr: sc.leaseMgr}
 			// Use a leased table descriptor for the backfill.
-			defer func() {
-				if err := tc.releaseTables(ctx, dontBlockForDBCacheUpdate); err != nil {
-					log.Warningf(ctx, "error releasing tables: %s", err)
-				}
-			}()
+			defer tc.releaseTables(ctx)
 			tableDesc, err := sc.getTableVersion(ctx, txn, tc, version)
 			if err != nil {
 				return err

--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -706,7 +706,7 @@ func (ex *connExecutor) close(ctx context.Context, closeType closeType) {
 		ex.state.finishExternalTxn()
 	}
 
-	if err := ex.resetExtraTxnState(ctx, txnAborted, ex.server.dbCache); err != nil {
+	if err := ex.resetExtraTxnState(ctx, ex.server.dbCache); err != nil {
 		log.Warningf(ctx, "error while cleaning up connExecutor: %s", err)
 	}
 
@@ -977,19 +977,12 @@ func (ns *prepStmtNamespace) copy() prepStmtNamespace {
 }
 
 func (ex *connExecutor) resetExtraTxnState(
-	ctx context.Context, ev txnEvent, dbCacheHolder *databaseCacheHolder,
+	ctx context.Context, dbCacheHolder *databaseCacheHolder,
 ) error {
 	ex.extraTxnState.schemaChangers.reset()
 
-	var opt releaseOpt
-	if ev == txnCommit {
-		opt = blockForDBCacheUpdate
-	} else {
-		opt = dontBlockForDBCacheUpdate
-	}
-	if err := ex.extraTxnState.tables.releaseTables(ctx, opt); err != nil {
-		return err
-	}
+	ex.extraTxnState.tables.releaseTables(ctx)
+
 	ex.extraTxnState.tables.databaseCache = dbCacheHolder.getDatabaseCache()
 
 	ex.extraTxnState.autoRetryCounter = 0
@@ -1846,11 +1839,6 @@ func (ex *connExecutor) txnStateTransitionsApplyWrapper(
 	case noEvent:
 	case txnStart:
 	case txnCommit:
-		// If we have schema changers to run, release leases early so that schema
-		// changers can run.
-		if len(ex.extraTxnState.schemaChangers.schemaChangers) > 0 {
-			ex.extraTxnState.tables.releaseLeases(ex.Ctx())
-		}
 		if schemaChangeErr := ex.extraTxnState.schemaChangers.execSchemaChanges(
 			ex.Ctx(), ex.server.cfg, &ex.sessionTracing,
 		); schemaChangeErr != nil {
@@ -1863,13 +1851,17 @@ func (ex *connExecutor) txnStateTransitionsApplyWrapper(
 			// schema changes failed. We don't have a good way to indicate this.
 			if implicitTxn {
 				res.SetError(schemaChangeErr)
-				return advInfo, nil
+			} else {
+				res.SetError(sqlbase.NewStatementCompletionUnknownError(schemaChangeErr))
 			}
-			res.SetError(sqlbase.NewStatementCompletionUnknownError(schemaChangeErr))
 		}
+
+		// Wait for the cache to reflect the dropped databases if any.
+		ex.extraTxnState.tables.waitForCacheToDropDatabases(ctx)
+
 		fallthrough
 	case txnRestart, txnAborted:
-		if err := ex.resetExtraTxnState(ex.Ctx(), advInfo.txnEvent, ex.server.dbCache); err != nil {
+		if err := ex.resetExtraTxnState(ex.Ctx(), ex.server.dbCache); err != nil {
 			return advanceInfo{}, err
 		}
 	default:

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -479,6 +479,30 @@ func (ex *connExecutor) checkTableTwoVersionInvariant(ctx context.Context) error
 	if txn.IsCommitted() {
 		panic("transaction has already committed")
 	}
+
+	// Release leases here for two reasons:
+	// 1. If there are existing leases at version V-2 for a descriptor
+	// being modified to version V being held the wait loop below that
+	// waits on a cluster wide release of old version leases will hang
+	// until these leases expire.
+	// 2. Once this transaction commits, the schema changers run and
+	// increment the version of the modified descriptors. If one of the
+	// descriptors being modified has a lease being held the schema
+	// changers will stall until the leases expire.
+	//
+	// The above two cases can be satified by releasing leases for both
+	// cases explicitly, but we prefer to call it here and kill two birds
+	// with one stone.
+	//
+	// It is safe to release leases even though the transaction hasn't yet
+	// committed only because the transaction timestamp has been fixed using
+	// CommitTimestamp().
+	//
+	// releaseLeases can fail to release a lease if the server is shutting
+	// down. This is okay because it will result in the two cases mentioned
+	// above simply hanging until the expiration time for the leases.
+	ex.extraTxnState.tables.releaseLeases(ex.Ctx())
+
 	count, err := CountLeases(ctx, ex.server.cfg.InternalExecutor, tables, txn.OrigTimestamp())
 	if err != nil {
 		return err
@@ -509,10 +533,6 @@ func (ex *connExecutor) checkTableTwoVersionInvariant(ctx context.Context) error
 	// We cleanup the transaction and create a new transaction wait time
 	// might be extensive and so we'd better get rid of all the intents.
 	txn.CleanupOnError(ctx, retryErr)
-
-	// Release leases held by the current transaction before waiting
-	// on cluster wide releases of old version leases.
-	ex.extraTxnState.tables.releaseLeases(ex.Ctx())
 
 	// Wait until all older version leases have been released or expired.
 	for r := retry.StartWithCtx(ctx, base.DefaultRetryOptions()); r.Next(); {

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -479,6 +479,9 @@ func (ex *connExecutor) checkTableTwoVersionInvariant(ctx context.Context) error
 	if txn.IsCommitted() {
 		panic("transaction has already committed")
 	}
+	if !txn.CommitTimestampFixed() {
+		panic("commit timestamp was not fixed")
+	}
 
 	// Release leases here for two reasons:
 	// 1. If there are existing leases at version V-2 for a descriptor

--- a/pkg/sql/exec_util.go
+++ b/pkg/sql/exec_util.go
@@ -380,7 +380,9 @@ type DistSQLPlannerTestingKnobs struct {
 // mutable cache, but nobody's sharing this holder. We should make up our mind
 // about whether we like the sharing or not and, if we do, share the holder too.
 // Also, we could use the SystemConfigDeltaFilter to limit the updates to
-// databases that chaged.
+// databases that chaged. One of the problems with the existing architecture
+// is if a transaction is completed on a session and the session remains dormant
+// for a long time, the next transaction will see a rather old database cache.
 type databaseCacheHolder struct {
 	mu struct {
 		syncutil.Mutex
@@ -403,20 +405,12 @@ func (dc *databaseCacheHolder) getDatabaseCache() *databaseCache {
 }
 
 // waitForCacheState implements the dbCacheSubscriber interface.
-func (dc *databaseCacheHolder) waitForCacheState(cond func(*databaseCache) (bool, error)) error {
+func (dc *databaseCacheHolder) waitForCacheState(cond func(*databaseCache) bool) {
 	dc.mu.Lock()
 	defer dc.mu.Unlock()
-	for {
-		done, err := cond(dc.mu.c)
-		if err != nil {
-			return err
-		}
-		if done {
-			break
-		}
+	for done := cond(dc.mu.c); !done; done = cond(dc.mu.c) {
 		dc.mu.cv.Wait()
 	}
-	return nil
 }
 
 // databaseCacheHolder implements the dbCacheSubscriber interface.

--- a/pkg/sql/internal.go
+++ b/pkg/sql/internal.go
@@ -203,7 +203,7 @@ func (ie *internalExecutorImpl) initConnEx(
 		// txn committed. But if there is a higher-level txn, they don't get
 		// released.
 		defer func() {
-			if err := ex.resetExtraTxnState(ctx, txnAborted, ex.server.dbCache); err != nil {
+			if err := ex.resetExtraTxnState(ctx, ex.server.dbCache); err != nil {
 				log.Warningf(ctx, "error while cleaning up connExecutor: %s", err)
 			}
 		}()

--- a/pkg/sql/lease.go
+++ b/pkg/sql/lease.go
@@ -215,7 +215,10 @@ func (s LeaseStore) acquire(ctx context.Context, tableID sqlbase.ID) (*tableVers
 	return table, err
 }
 
-// Release a previously acquired table descriptor.
+// Release a previously acquired table descriptor. Never let this method
+// read a table descriptor because it can be called while modifying a
+// descriptor through a schema change before the schema change has committed
+// that can result in a deadlock.
 func (s LeaseStore) release(ctx context.Context, stopper *stop.Stopper, table *tableVersionState) {
 	retryOptions := base.DefaultRetryOptions()
 	retryOptions.Closer = stopper.ShouldQuiesce()
@@ -320,6 +323,10 @@ func incrementVersion(
 	// the mvcc timestamp of the descriptor. This requires moving the
 	// schema change lease out of the descriptor making the
 	// descriptor truly immutable at a version.
+	// Also recognize that the leases are released before the transaction
+	// is committed through a call to TableCollection.releaseLeases(),
+	// so updating this policy will also need to consider not doing
+	// that.
 	modTime := txn.CommitTimestamp()
 	tableDesc.ModificationTime = modTime
 	log.Infof(ctx, "publish: descID=%d (%s) version=%d mtime=%s",
@@ -984,7 +991,7 @@ func (t *tableState) release(
 	return nil, nil
 }
 
-// release the lease associated with the table version.
+// releaseLease associated with the table version.
 func releaseLease(table *tableVersionState, m *LeaseManager) {
 	m.tableNames.remove(table)
 

--- a/pkg/sql/schema_changer.go
+++ b/pkg/sql/schema_changer.go
@@ -197,7 +197,7 @@ func (e errTableVersionMismatch) Error() string {
 // AcquireLease acquires a schema change lease on the table if
 // an unexpired lease doesn't exist. It returns the lease.
 func (sc *SchemaChanger) AcquireLease(
-	ctx context.Context, tables *TableCollection,
+	ctx context.Context,
 ) (sqlbase.TableDescriptor_SchemaChangeLease, error) {
 	var lease sqlbase.TableDescriptor_SchemaChangeLease
 	err := sc.db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
@@ -224,13 +224,6 @@ func (sc *SchemaChanger) AcquireLease(
 		}
 		lease = sc.createSchemaChangeLease()
 		tableDesc.Lease = &lease
-		if tables != nil {
-			// tables is nil when running schema changes from the background schema change task.
-			// When running schema changes at the end of a SQL txn, tables is the
-			// table collection used as descriptor cache by queries. That must
-			// remain consistent with the KV state, so we must update it here.
-			tables.addUncommittedTable(*tableDesc)
-		}
 		return txn.Put(ctx, sqlbase.MakeDescMetadataKey(tableDesc.ID), sqlbase.WrapDescriptor(tableDesc))
 	})
 	return lease, err
@@ -255,7 +248,7 @@ func (sc *SchemaChanger) findTableWithLease(
 // ReleaseLease releases the table lease if it is the one registered with
 // the table descriptor.
 func (sc *SchemaChanger) ReleaseLease(
-	ctx context.Context, lease sqlbase.TableDescriptor_SchemaChangeLease, tables *TableCollection,
+	ctx context.Context, lease sqlbase.TableDescriptor_SchemaChangeLease,
 ) error {
 	return sc.db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
 		tableDesc, err := sc.findTableWithLease(ctx, txn, lease)
@@ -265,13 +258,6 @@ func (sc *SchemaChanger) ReleaseLease(
 		tableDesc.Lease = nil
 		if err := txn.SetSystemConfigTrigger(); err != nil {
 			return err
-		}
-		if tables != nil {
-			// tables is nil when running schema changes from the background schema change task.
-			// When running schema changes at the end of a SQL txn, tables is the
-			// table collection used as descriptor cache by queries. That must
-			// remain consistent with the KV state, so we must update it here.
-			tables.addUncommittedTable(*tableDesc)
 		}
 		return txn.Put(ctx, sqlbase.MakeDescMetadataKey(tableDesc.ID), sqlbase.WrapDescriptor(tableDesc))
 	})
@@ -552,7 +538,7 @@ func (sc *SchemaChanger) exec(
 			sc.tableID, sc.mutationID)
 	}
 	// Acquire lease.
-	lease, err := sc.AcquireLease(ctx, evalCtx.Tables)
+	lease, err := sc.AcquireLease(ctx)
 	if err != nil {
 		return err
 	}
@@ -564,7 +550,7 @@ func (sc *SchemaChanger) exec(
 		if !needRelease {
 			return
 		}
-		if err := sc.ReleaseLease(ctx, lease, evalCtx.Tables); err != nil {
+		if err := sc.ReleaseLease(ctx, lease); err != nil {
 			log.Warning(ctx, err)
 		}
 	}()

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -93,7 +93,7 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 	ctx := context.TODO()
 
 	// Acquire a lease.
-	lease, err := changer.AcquireLease(ctx, nil)
+	lease, err := changer.AcquireLease(ctx)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -104,7 +104,7 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 	}
 
 	// Acquiring another lease will fail.
-	if _, err := changer.AcquireLease(ctx, nil); !testutils.IsError(
+	if _, err := changer.AcquireLease(ctx); !testutils.IsError(
 		err, "an outstanding schema change lease exists",
 	) {
 		t.Fatal(err)
@@ -131,12 +131,12 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 	}
 
 	// Releasing an old lease fails.
-	if err := changer.ReleaseLease(ctx, oldLease, nil); err == nil {
+	if err := changer.ReleaseLease(ctx, oldLease); err == nil {
 		t.Fatal("releasing a old lease succeeded")
 	}
 
 	// Release lease.
-	if err := changer.ReleaseLease(ctx, lease, nil); err != nil {
+	if err := changer.ReleaseLease(ctx, lease); err != nil {
 		t.Fatal(err)
 	}
 
@@ -146,7 +146,7 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 	}
 
 	// acquiring the lease succeeds
-	lease, err = changer.AcquireLease(ctx, nil)
+	lease, err = changer.AcquireLease(ctx)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -480,7 +480,7 @@ func runSchemaChangeWithOperations(
 	// Grabbing a schema change lease on the table will fail, disallowing
 	// another schema change from being simultaneously executed.
 	sc := sql.NewSchemaChangerForTesting(tableDesc.ID, 0, 0, *kvDB, nil, jobRegistry, execCfg)
-	if l, err := sc.AcquireLease(ctx, nil); err == nil {
+	if l, err := sc.AcquireLease(ctx); err == nil {
 		t.Fatalf("schema change lease acquisition on table %d succeeded: %v", tableDesc.ID, l)
 	}
 

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -2090,6 +2090,113 @@ CREATE TABLE t.test (
 	}
 }
 
+// Test a DROP failure on a unique column. The rollback
+// process might not be able to reconstruct the index and thus
+// recreates the column as non-UNIQUE. For now this is considered
+// acceptable.
+func TestSchemaUniqueColumnDropFailure(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	params, _ := tests.CreateTestServerParams()
+	const chunkSize = 200
+	attempts := 0
+	// DROP UNIQUE COLUMN is executed in two steps: drop index and drop column.
+	// Chunked backfill attempts:
+	// attempt 1-5: drop index
+	// attempt 6-7: drop part of the column before hitting a schema
+	// change error.
+	// purge the schema change.
+	const expectedAttempts = 7
+	const maxValue = (expectedAttempts/2+1)*chunkSize + 1
+	params.Knobs = base.TestingKnobs{
+		SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
+			AsyncExecNotification: asyncSchemaChangerDisabled,
+			BackfillChunkSize:     chunkSize,
+			// Aggressively checkpoint, so that a schema change
+			// failure happens after a checkpoint has been written.
+			WriteCheckpointInterval: time.Nanosecond,
+		},
+		DistSQL: &distsqlrun.TestingKnobs{
+			RunBeforeBackfillChunk: func(sp roachpb.Span) error {
+				attempts++
+				// Return a deadline exceeded error while dropping
+				// the column after the index has been dropped.
+				if attempts == expectedAttempts {
+					return errors.New("permanent failure")
+				}
+				return nil
+			},
+		},
+		// Disable backfill migrations so it doesn't interfere with the
+		// backfill in this test.
+		SQLMigrationManager: &sqlmigrations.MigrationManagerTestingKnobs{
+			DisableBackfillMigrations: true,
+		},
+	}
+	server, sqlDB, kvDB := serverutils.StartServer(t, params)
+	defer server.Stopper().Stop(context.TODO())
+
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE t;
+CREATE TABLE t.test (k INT PRIMARY KEY, v INT UNIQUE DEFAULT 23 CREATE FAMILY F3);
+`); err != nil {
+		t.Fatal(err)
+	}
+
+	// Bulk insert.
+	if err := bulkInsertIntoTable(sqlDB, maxValue); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := checkTableKeyCount(context.TODO(), kvDB, 2, maxValue); err != nil {
+		t.Fatal(err)
+	}
+
+	// A schema change that fails.
+	if _, err := sqlDB.Exec(`ALTER TABLE t.test DROP column v`); !testutils.IsError(err, `permanent failure`) {
+		t.Fatalf("err = %s", err)
+	}
+
+	// The index is not regenerated.
+	if err := checkTableKeyCount(context.TODO(), kvDB, 1, maxValue); err != nil {
+		t.Fatal(err)
+	}
+
+	// Column v still exists with the default value.
+	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
+	if e := 2; e != len(tableDesc.Columns) {
+		t.Fatalf("e = %d, v = %d, columns = %+v", e, len(tableDesc.Columns), tableDesc.Columns)
+	} else if tableDesc.Columns[0].Name != "k" || tableDesc.Columns[1].Name != "v" {
+		t.Fatalf("columns %+v", tableDesc.Columns)
+	} else if len(tableDesc.Mutations) > 0 {
+		t.Fatalf("mutations %+v", tableDesc.Mutations)
+	}
+
+	rows, err := sqlDB.Query(`SELECT v from t.test`)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer rows.Close()
+
+	count := 0
+	for ; rows.Next(); count++ {
+		var v int
+		if err := rows.Scan(&v); err != nil {
+			t.Errorf("row %d scan failed: %s", count, err)
+			continue
+		}
+		if 23 != v {
+			t.Errorf("e = %d, v = %d", 23, v)
+		}
+	}
+	if err := rows.Err(); err != nil {
+		t.Fatal(err)
+	}
+	eCount := maxValue + 1
+	if eCount != count {
+		t.Fatalf("read the wrong number of rows: e = %d, v = %d", eCount, count)
+	}
+}
+
 // Test an UPDATE using a primary and a secondary index in the middle
 // of a column backfill.
 func TestUpdateDuringColumnBackfill(t *testing.T) {

--- a/pkg/sql/table.go
+++ b/pkg/sql/table.go
@@ -165,8 +165,6 @@ type TableCollection struct {
 
 	// dbCacheSubscriber is used to block until the node's database cache has been
 	// updated when releaseTables is called.
-	// Can be nil if releaseTables() is only called with the
-	// dontBlockForDBCacheUpdate option.
 	dbCacheSubscriber dbCacheSubscriber
 
 	// Same as uncommittedTables applying to databases modified within
@@ -185,7 +183,7 @@ type dbCacheSubscriber interface {
 	// waitForCacheState takes a callback depending on the cache state and blocks
 	// until the callback declares success. The callback is repeatedly called as
 	// the cache is updated.
-	waitForCacheState(cond func(*databaseCache) (bool, error)) error
+	waitForCacheState(cond func(*databaseCache) bool)
 }
 
 // getTableVersion returns a table descriptor with a version suitable for
@@ -367,19 +365,6 @@ func (tc *TableCollection) getTableVersionByID(
 	return table, nil
 }
 
-// releaseOpt specifies options for tc.releaseTables().
-type releaseOpt bool
-
-const (
-	// blockForDBCacheUpdate makes releaseTables() block until the node's database
-	// cache has been updated to reflect the dropped or renamed databases. If used
-	// within a SQL session, this ensures that future queries on that session
-	// behave correctly when trying to use the names of the recently
-	// dropped/renamed databases.
-	blockForDBCacheUpdate     releaseOpt = true
-	dontBlockForDBCacheUpdate releaseOpt = false
-)
-
 func (tc *TableCollection) releaseLeases(ctx context.Context) {
 	if len(tc.leasedTables) > 0 {
 		log.VEventf(ctx, 2, "releasing %d tables", len(tc.leasedTables))
@@ -393,52 +378,46 @@ func (tc *TableCollection) releaseLeases(ctx context.Context) {
 }
 
 // releaseTables releases all tables currently held by the TableCollection.
-func (tc *TableCollection) releaseTables(ctx context.Context, opt releaseOpt) error {
-	if len(tc.leasedTables) > 0 {
-		log.VEventf(ctx, 2, "releasing %d tables", len(tc.leasedTables))
-		for _, table := range tc.leasedTables {
-			if err := tc.leaseMgr.Release(table); err != nil {
-				log.Warning(ctx, err)
-			}
-		}
-		tc.leasedTables = tc.leasedTables[:0]
-	}
+func (tc *TableCollection) releaseTables(ctx context.Context) {
+	tc.releaseLeases(ctx)
 	tc.uncommittedTables = nil
 	tc.createdTables = nil
-
-	if opt == blockForDBCacheUpdate {
-		for _, uc := range tc.uncommittedDatabases {
-			if !uc.dropped {
-				continue
-			}
-			// Wait until the database cache has been updated to properly
-			// reflect a dropped database, so that future commands on the
-			// same gateway node observe the dropped database.
-			err := tc.dbCacheSubscriber.waitForCacheState(
-				func(dc *databaseCache) (bool, error) {
-					// Resolve the database name from the database cache.
-					dbID, err := dc.getDatabaseID(ctx,
-						tc.leaseMgr.execCfg.DB.Txn, uc.name, false /*required*/)
-					if err != nil || dbID == 0 {
-						// dbID can still be 0 if required is false and
-						// the database is not found.
-						return true, err
-					}
-
-					// If the database name still exists but it now references another
-					// db with a more recent id, we're good - it means that the database
-					// name has been reused.
-					return dbID > uc.id, nil
-				})
-			if err != nil {
-				return err
-			}
-		}
-	}
-
 	tc.uncommittedDatabases = nil
 	tc.releaseAllDescriptors()
-	return nil
+}
+
+// Wait until the database cache has been updated to properly
+// reflect all dropped databases, so that future commands on the
+// same gateway node observe the dropped databases.
+func (tc *TableCollection) waitForCacheToDropDatabases(ctx context.Context) {
+	for _, uc := range tc.uncommittedDatabases {
+		if !uc.dropped {
+			continue
+		}
+		// Wait until the database cache has been updated to properly
+		// reflect a dropped database, so that future commands on the
+		// same gateway node observe the dropped database.
+		tc.dbCacheSubscriber.waitForCacheState(
+			func(dc *databaseCache) bool {
+				// Resolve the database name from the database cache.
+				dbID, err := dc.getDatabaseID(ctx,
+					tc.leaseMgr.execCfg.DB.Txn, uc.name, false /*required*/)
+				if err != nil || dbID == 0 {
+					// dbID can still be 0 if required is false and
+					// the database is not found. Swallowing error here
+					// because it was felt there was no value in returning
+					// it to a higher layer only to be swallow there. This
+					// entire codepath is only called from one place so
+					// it's better to swallow it here.
+					return true
+				}
+
+				// If the database name still exists but it now references another
+				// db with a more recent id, we're good - it means that the database
+				// name has been reused.
+				return dbID > uc.id
+			})
+	}
 }
 
 func (tc *TableCollection) addUncommittedTable(desc sqlbase.TableDescriptor) {

